opentelemetry/src/sdk/trace/span_processor.rs

```diff
@@ -547,4 +670,108 @@ mod tests {
             "timed out in 5 seconds. force_flush may not export any data when called"
         );
     }
+
+    struct BlockingExporter<D> {
+        delay_for: time::Duration,
+        delay_fn: D,
+    }
+
+    impl<D, DS> Debug for BlockingExporter<D>
+    where
+        D: Fn(time::Duration) -> DS + 'static + Send + Sync,
+        DS: Future<Output = ()> + Send + Sync + 'static,
+    {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            f.write_str("blocking exporter for testing")
+        }
+    }
+
+    #[async_trait]
+    impl<D, DS> SpanExporter for BlockingExporter<D>
+    where
+        D: Fn(time::Duration) -> DS + 'static + Send + Sync,
+        DS: Future<Output = ()> + Send + Sync + 'static,
+    {
+        async fn export(&mut self, batch: Vec<SpanData>) -> ExportResult {
+            println!("Accepting {} spans", batch.len());
+            (self.delay_fn)(self.delay_for).await;
+            println!("Finish exporting, return result from exporter");
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_timeout() {
+        // If time_out is true, then we ask exporter to block for 60s and set timeout to 5s.
+        // If time_out is false, then we ask the exporter to block for 5s and set timeout to 60s.
+        // Either way, the test should be finished within 5s.
+        let mut runtime = tokio::runtime::Builder::new()
+            .threaded_scheduler()
+            .enable_all()
+            .build()
+            .unwrap();
+        runtime.block_on(timeout_test_tokio(true));
+        runtime.block_on(timeout_test_tokio(false));
+
+        async_std::task::block_on(timeout_test_std_async(true));
+        async_std::task::block_on(timeout_test_std_async(false));
```

Comment:
It might be nice to make 4 different tests out of those. This makes it easier to figure out what went wrong if one of them fails.